shared/js/background/chrome-mv3-redirect.es6.js

```diff
@@ -0,0 +1,210 @@
+        // just default to true.
+        const sameDomain = isSameDomainRequest(thisTab, requestData)
+
+        // only count trackers on pages with 200 response. Trackers on these sites are still
+        // blocked below but not counted on the popup. We can also run into a case where
+        // we block a tracker faster then we can update the tab so we check sameDomain.
+        if (thisTab.statusCode === 200 && sameDomain) {
+            // record all tracker urls on a site even if we don't block them
+            thisTab.site.addTracker(tracker)
+
+            // record potential blocked trackers for this tab
+            thisTab.addToTrackers(tracker)
+        }
+
+        browserWrapper.notifyPopup({'updateTabData': true})
+
+        // Block the request if the site is not whitelisted
+        if (!thisTab.site.whitelisted && tracker.action.match(/block|redirect/)) {
+            if (sameDomain) thisTab.addOrUpdateTrackersBlocked(tracker)
+
+            // update badge icon for any requests that come in after
+            // the tab has finished loading
+            if (thisTab.status === 'complete') thisTab.updateBadgeIcon()
+
+            if (thisTab.statusCode === 200) {
+                Companies.add(tracker.tracker.owner)
+            }
+
+            // Initiate hiding of blocked ad DOM elements
+            tryElementHide(requestData, thisTab)
+
+            console.info(`blocked on ${utils.extractHostFromURL(thisTab.url)}: [${tracker.tracker.owner.name}] ${requestData.url}`)
+        }
+    }
+}
+
+/**
+ * Make sure that HTTPS upgrade done by declarativeNetRequest rules was valid
+ */
+function verifyRedirect (requestData) {
+    const { tabId, url, redirectUrl, statusCode } = requestData
+
+    // We are only interested in internal redirects (done by the extension)
+    if (statusCode !== 307) {
+        return
+    }
+
+    let fromUrl, toUrl
+    try {
+        fromUrl = new URL(url)
+        toUrl = new URL(redirectUrl)
+    } catch (e) {
+        // Ingore invalid URLs
+        return
+    }
+
+    // Ignore if redirect is not from HTTP to HTTPS
+    if (fromUrl.protocol !== 'http:' || toUrl.protocol !== 'https:') {
+        return
+    }
+
+    const upgradedFromUrl = new URL(fromUrl)
+    upgradedFromUrl.protocol = 'https:'
+
+    // Ignore if redirect changed anything about the request other than the protocol
+    if (upgradedFromUrl.toString() !== toUrl.toString()) {
+        return
+    }
+
+    console.info(`Checking if ${redirectUrl} can continue over HTTPS.`)
+
+    // get information about the tab
+    const tabManagerTab = tabManager.get(requestData)
+    const isMainFrame = requestData.type === 'main_frame'
+    let isUpgradable = https.canUpgradeUrl(requestData.url)
+
+    // if returned result is an boolean, and not a promise, put it into a resolved promise for easier handling
+    if (!(isUpgradable instanceof Promise)) {
+        isUpgradable = Promise.resolve(isUpgradable)
+    }
```

Comment:
Feels like we should get `canUpgradeUrl` to always return the same value but ah well. This works for now.